lattices/src/test.rs

```diff
@@ -0,0 +1,196 @@
+
+    // PartialEq:
+    // a != b if and only if !(a == b).
+    cartesian_power(items).for_each(|[a, b]| assert_eq!(a != b, !(a == b)));
+
+    // Eq:
+    // reflexive: a == a;
+    items.iter().for_each(|a| assert!(a == a));
+    // symmetric: a == b implies b == a; and
+    cartesian_power(items).for_each(|[a, b]| assert_eq!(a == b, b == a));
+    // transitive: a == b and b == c implies a == c.
+    cartesian_power(items).for_each(|[a, b, c]| {
+        if a == b && b == c {
+            assert_eq!(a == b && b == c, a == c);
+        }
+    });
+
+    // PartialOrd
+    cartesian_power(items).for_each(|[a, b]| {
+        // a == b if and only if partial_cmp(a, b) == Some(Equal).
+        assert_eq!(a == b, a.partial_cmp(b) == Some(Equal));
+        // a < b if and only if partial_cmp(a, b) == Some(Less)
+        assert_eq!(a < b, a.partial_cmp(b) == Some(Less));
+        // a > b if and only if partial_cmp(a, b) == Some(Greater)
+        assert_eq!(a > b, a.partial_cmp(b) == Some(Greater));
+        // a <= b if and only if a < b || a == b
+        assert_eq!(a <= b, a < b || a == b);
+        // a >= b if and only if a > b || a == b
+        assert_eq!(a >= b, a > b || a == b);
+        // a != b if and only if !(a == b).
+        assert_eq!(a != b, !(a == b));
+    });
+    // transitivity: a < b and b < c implies a < c. The same must hold for both == and >.
+    cartesian_power(items).for_each(|[a, b, c]| {
+        if a < b && b < c {
+            assert!(a < c);
+        }
+        if a == b && b == c {
+            assert!(a == c);
+        }
+        if a > b && b > c {
+            assert!(a > c);
+        }
+    });
+    // duality: a < b if and only if b > a.
+    cartesian_power(items).for_each(|[a, b]| assert_eq!(a < b, b > a));
+}
+
+/// Check lattice associativity, commutativity, and idempotence.
+pub fn check_lattice_properties<T: Merge<T> + Clone + Eq + Debug>(items: &[T]) {
+    // Idempotency
+    // x ∧ x = x
+    items
+        .iter()
+        .for_each(|x| assert_eq!(Merge::merge_owned(x.to_owned(), x.to_owned()), x.to_owned()));
+
+    // Commutativity
+    // x ∧ y = y ∧ x
+    cartesian_power(items).for_each(|[x, y]| {
+        assert_eq!(
+            Merge::merge_owned(x.to_owned(), y.to_owned()),
+            Merge::merge_owned(y.to_owned(), x.to_owned())
+        )
+    });
+
+    // Associativity
+    // x ∧ (y ∧ z) = (x ∧ y) ∧ z
+    cartesian_power(items).for_each(|[x, y, z]| {
+        assert_eq!(
+            Merge::merge_owned(x.to_owned(), Merge::merge_owned(y.to_owned(), z.to_owned())),
+            Merge::merge_owned(Merge::merge_owned(x.to_owned(), y.to_owned()), z.to_owned())
+        )
+    });
+}
+
+/// Returns an iterator of `N`-length arrays containing all possible permutations (with
+/// replacement) of items in `items`. I.e. the `N`th cartesian power of `items`. I.e. the cartesian
+/// product of `items` with itself `N` times.
+pub fn cartesian_power<T, const N: usize>(
```

Comment:
Would something like this work? https://docs.rs/itertools/0.8.2/itertools/trait.Itertools.html#method.permutations

Reply:
Would be [multi_cartesian_product](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.multi_cartesian_product) or [iproduct](https://docs.rs/itertools/latest/itertools/macro.iproduct.html#) but this is more ergonomic w/ `const N` and avoids adding a dependency to cargo.toml